### ISSUE TITLE
Support running the lwaftr over virtio interfaces

### ISSUE THREAD

**Issue by wingo** [Igalia/snabbswitch#199]
We would like to be able to run the lwaftr under virtualized network interfaces, using the virtio-net implementation from https://github.com/SnabbCo/snabbswitch/pull/681.

There are a few subtasks here.

 1. Merging.  Right now we are based off the rambutan release, which currently corresponds to Snabb master.  The PR in question is based on Snabb master, so we should be able to merge right now without conflicts and improve as we go.
 2. Performance.  Currently on a simple forwarding app we are seeing 50% of time being spent in the interpreter, and although this app is able to do 4MPPS Gbps for 500-byte packets, it's not able to do line rate for smaller packets.  We should re-test with the latest code as of yesterday and see if we still see the issue, and if so come up with a fix and possibly get it into the initial PR.
 3. Configuration.  Right now the `snabb-lwaftr run` command will always try to access Intel cards.  We need it to use virtio cards if needed.
 4. Testing.  How to set up a test environment?  It's pretty complicated unfortunately.  Ideally someone should be able to run a benchmark using one command.

This is a big task.  I suggest we go ahead and merge the branch for the PR and work on performance now.  Ideally we can find the perf problem today or tomorrow and submit upstream.  With a bit of luck this change will lead to excellent perf and all of our worries will go away!

**Comment by dpino** [comment]
The plan sounds good to me. I can take care of most these subtasks.

With regard to 1), the PR can be merged without changes I think. Let me test the PR on top lwaftr_rambutan first to confirm (working on that now).

With regard to performance, I have been trying to fix the code and make it run fully compiled, but still not success. I can share my investigations and progress here.

With regard to configuration, that should be easy. I can push a PR with that change.

With regard to testing, I think it's possible to have one bash script that can benchmark the app with the virtio settings automatically. But a VM image is needed, in any case.

**Comment by dpino** [comment]
#202 solves checkpoint 1). It's the upstream VirtioNet PR rebased to `lwaftr_rambutan`.

**Comment by dpino** [comment]
#203 solves checkpoint 3). Adds flag '--virtio_net' or '-i' (short form), so virtio_net drivers are used when launching `snabb-lwaftr run`.

**Comment by dpino** [comment]
#206 solves checkpoint 4). 

The bad performance observed in the benchmarking of the virtio_net driver was due to a misconfiguration of the virtualized NICs in the host. Instead of using snabbnfv, I was using a script that brought up two NICs with vhost-user, that means two NICs where handled by one core. We want each card to be handled by a different NIC. When that happens, performance is the same as the results we got running the lwaftr in baremetal (really good news).

In conclusion, checkpoint 2) is now solved too.

**Comment by kbara** [comment]
Awesome work.

**Comment by dpino** [comment]
Closing as all the checkpoints in this issue have been addressed. #230 adds a script to easily deploy the lwAFTR virtualized.

**Comment by kbara** [comment]
Awesome work, @dpino !
